Yifanzhou/add migration guide (#43176)
Azure/azure-sdk-for-net: sdk/purview/Azure.Analytics.Purview.DataMap/MigrationGuide.md

# Migrate from Purview Catalog to Purview DataMap

This guide is intended to assist in the migration to Azure Purview DataMap client library [`Azure.Analytics.Purview.DataMap`](https://www.nuget.org/packages/Azure.Analytics.Purview.DataMap) from [`Azure.Analytics.Purview.Catalog`](https://www.nuget.org/packages/Azure.Analytics.Purview.Catalog). It will focus on side-by-side comparisons for similar operations between the two packages.

> For those new to the Purview DataMAP library for .NET, please refer to the [`Azure.Analytics.Purview.DataMap` README](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/purview/Azure.Analytics.Purview.DataMap/README.md) and [`Azure.Analytics.Purview.DataMap` samples](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/purview/Azure.Analytics.Purview.DataMap/samples) for the `Azure.Analytics.Purview.DataMap` library rather than this guide.

Reviewer: ```suggestion
For those new to the Purview DataMap library for .NET, please refer to the [`Azure.Analytics.Purview.DataMap` README](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/purview/Azure.Analytics.Purview.DataMap/README.md) and [`Azure.Analytics.Purview.DataMap` samples](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/purview/Azure.Analytics.Purview.DataMap/samples) for the `Azure.Analytics.Purview.DataMap` library rather than this guide.
```

The other uses of "DataMap" in this document do not use the capitalization here.   I'm assuming this was a typo.

Author: Updated. thanks

There were several areas of consistent feedback expressed across the Azure client library ecosystem. One of the most important is that the client libraries for different Azure services have not had a consistent approach to organization, naming, and API structure. Additionally, many developers have felt that the learning curve was difficult, and the APIs did not offer a good, approachable, and consistent onboarding story for those learning Azure or exploring a specific Azure service.

To try and improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [.NET-specific guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html) was also introduced to ensure that .NET clients have a natural and idiomatic feel that mirrors that of the .NET base class libraries. Further details are available in the guidelines for those interested.

> The new Purview DataMap library `Azure.Analytics.Purview.DataMap` includes the service models together with the GA DataMap APIs [API Document](https://learn.microsoft.com/rest/api/purview/datamapdataplane/operation-groups?view=rest-purview-datamapdataplane-2023-09-01). The client name and the operation names have slightly changed but the main functionality remains the same.

Reviewer: ```suggestion
The new Purview DataMap library `Azure.Analytics.Purview.DataMap` includes the service models together with the GA DataMap APIs [API Document](hhttps://learn.microsoft.com/rest/api/purview/datamapdataplane/operation-groups). The client name and the operation names have slightly changed but the main functionality remains the same.
```

nit:  You don't want to link to a specific service version, or this document will become stale each time the service updates.

Author: Updated thanks.

## General changes

### Package and client name

Previously in `Azure.Analytics.Purview.Catalog`, the service client name is PurviewCatalogClient.

> ```C#

Reviewer: All code here must be [snippets](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#updating-sample-snippets), just like the library examples.

Reviewer: Catalog should not have been approved.   All samples in this repository are required to use snippets.

var credential = new DefaultAzureCredential();
var client = new PurviewCatalogClient(new Uri("https://<my-account-name>.purview.azure.com"), credential);
```

> Now in `Azure.Analytics.Purview.DataMap`, the service client name is DataMapClient.

Reviewer: Are all the operations the same on the new client?   Do they have the same names and take the same parameters?   If not, we'll want to identify the key champion scenarios for the library and show examples of how it used to look and how to perform that same operation in the new client.

Author: Sample added.

Reviewer: You've shown how to create clients, which is a good start.   Are all of the service operations the same and the client just has a different name?  

The goal for a migration guide is to walk customers through the key champion scenarios for using the library and provide examples for old and new.   We're doing that here with client creation, but nothing demonstrates the differences in how the clients are used for customer scenarios.

For an example, see:  [Event Hubs migration guide](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/MigrationGuide.md)

To try and improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [.NET-specific guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html) was also introduced to ensure that .NET clients have a natural and idiomatic feel that mirrors that of the .NET base class libraries. Further details are available in the guidelines for those interested.

The new Purview DataMap library `Azure.Analytics.Purview.DataMap` includes the service models together with the DataMap APIs [API Document](https://learn.microsoft.com/rest/api/purview/datamapdataplane/operation-groups). The client name and the operation names have slightly changed but the main functionality remains the same.

## General changes

### Package and client name

> Previously in `Azure.Analytics.Purview.Catalog`, the service client name is PurviewCatalogClient.

Reviewer: ```suggestion
Previously in `Azure.Analytics.Purview.Catalog`, the service client name is `PurviewCatalogClient`.
```

```C#
var credential = new DefaultAzureCredential();
var client = new PurviewCatalogClient(new Uri("https://<my-account-name>.purview.azure.com"), credential);
```

> Now in `Azure.Analytics.Purview.DataMap`, the service client name is DataMapClient.

Reviewer: ```suggestion
Now in `Azure.Analytics.Purview.DataMap`, the service client name is `DataMapClient`.
```

```C# Snippet:CreateDataMapClient

> Uri endpoint = TestEnvironment.Endpoint;

Reviewer: We shouldn't be using `TestEnvironment.Endpoint` here, as it is not a type that customers will have if they copy/paste.  Please update to show an example.  Something that is an obvious fake value like:

```CSharp
Uri endpoint = new Uri("<some-name>.service.azure.net");
```

If your service has a consistent pattern in the domain portion, it would be good to show it.
